server/src/main/java/org/elasticsearch/cluster/metadata/IndexAbstractionResolver.java

```diff
@@ -139,6 +157,22 @@ public static boolean isIndexVisible(String expression, String index, IndicesOpt
         if (isVisible == false) {
             return false;
         }
+        if (indexAbstraction.isSystem()) {
+            // system index that backs system data stream
+            if (indexAbstraction.getParentDataStream() != null) {
+                if (indexAbstraction.getParentDataStream().isSystem() == false) {
+                    throw new IllegalStateException("system index is part of a data stream that is not a system data stream");
```

Comment:
I think this is a worth candidate for promoting to an `assert` - we should never hit this case, right?

Reply:
it should never hit this case but I prefer this actually being detected in the wild since we don't run with assertions on normally

Reply:
++, I meant along the lines of:
```java
assert false : "this means you probably wrote a bug that lets system indices be part of non data streams";
throw new IllegalStateException("system index is part of a data stream that is not a system data stream"); 
```

This shouldn't block the PR though, more maybe as a follow-up?